ecllibrary/teststd/str/TestWeightedDamerauLevenshteinDistance.ecl

```diff
@@ -0,0 +1,49 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems®.  All rights reserved.
+############################################################################## */
+
+IMPORT Std.Str;
+
+EXPORT TestWeightedDamerauLevenshteinDistance := MODULE
+
+  STRING alpha := 'abcdefghijklmnopqrstuvwxyz';
+  STRING digits := '0123456789';
+  STRING largeAlpha := alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha;
+  STRING manyAlpha := alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha+alpha;
+  STRING manyDigits := digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+
+                       digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits+digits;
+
+  EXPORT TestConst := MODULE
+    EXPORT Test01 := ASSERT(Str.WeightedDamerauLevenshteinDistance('','') = 0, CONST);
+    EXPORT Test02 := ASSERT(Str.WeightedDamerauLevenshteinDistance('','                ') = 0, CONST);
+    EXPORT Test03 := ASSERT(Str.WeightedDamerauLevenshteinDistance('                ','') = 0, CONST);
+    EXPORT Test04 := ASSERT(Str.WeightedDamerauLevenshteinDistance('a ','                ') = 1, CONST);
+    //EXPORT Test05 := ASSERT(Str.WeightedDamerauLevenshteinDistance(' a ','   ') = 1, CONST);
+    EXPORT Test06 := ASSERT(Str.WeightedDamerauLevenshteinDistance('Aprs  ','APp') = 2, CONST);//default confusion matrix is case insensitive.
+    EXPORT Test07 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','acbd') = 1, CONST);
+    EXPORT Test08 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','abd') = 1, CONST);
+    EXPORT Test09 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','abc') = 1, CONST);
+    EXPORT Test10 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','bcd') = 1, CONST);
+    EXPORT Test11 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','abcde') = 1, CONST);
+    EXPORT Test12 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','aabcd') = 1, CONST);
+    EXPORT Test13 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd',' abcd') = 1, CONST);
+    EXPORT Test14 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','a bcd') = 1, CONST);
+    EXPORT Test15 := ASSERT(Str.WeightedDamerauLevenshteinDistance('abcd','adcd') = 0, CONST);
```

Comment:
It is due to the weight matrix in that 'b' & 'd' can be optically confused and thus have a zero weight.

So the other three distance metrics give a distance of 1 for these two strings and thus to differentiate with the weighted version the only option is for the operation to be ignored and thus yield 0.

It is a good reason to expand to allow user defined weight matrices.

More importantly thank you for noting this as I had forgotten about it and assumed, as you did, for the auto-id-suggestion code, that d = 0 => identical-identical rather than weighted-identical.